src/query/ts/m3db/encoded_step_iterator_generic.go

```diff
@@ -88,43 +101,103 @@ func (it *encodedStepIterWithCollector) nextForStep(
 		// If this datapoint is before the current timestamp, add it as a
 		// consolidation candidate.
 		if !dp.Timestamp.After(stepTime) {
-			it.seriesPeek[i].started = false
-			it.collector.AddPointForIterator(dp, i)
+			peek.started = false
+			collector.AddPoint(dp)
 		} else {
 			// This point exists further than the current step.
 			// Set peeked value to this point, then consolidate the retrieved
 			// series.
-			it.seriesPeek[i].point = dp
-			it.seriesPeek[i].started = true
-			return
+			peek.point = dp
+			peek.started = true
+			return peek, collector, nil
 		}
 	}
 
-	if err := iter.Err(); err != nil {
-		it.err = err
+	peek.finished = true
+	return peek, collector, iter.Err()
+}
+
+func (it *encodedStepIterWithCollector) nextParallel() error {
+	var (
+		stepTime = it.stepTime
+
+		multiErr xerrors.MultiError
+	)
+
+	it.wg.Add(len(it.seriesIters))
+	for i := range it.seriesIters {
+		var (
+			i         = i
+			peek      = it.seriesPeek[i]
+			iter      = it.seriesIters[i]
+			collector = it.seriesCollectors[i]
+		)
+
+		it.workerPool.Go(func() {
+			peek, collector, err := nextForStep(peek, iter, collector, stepTime)
+			it.mu.Lock()
+			it.seriesPeek[i] = peek
+			it.seriesCollectors[i] = collector
+			multiErr = multiErr.Add(err)
+			it.mu.Unlock()
+			it.wg.Done()
+		})
+	}
+
+	it.wg.Wait()
+	if it.err = multiErr.FinalError(); it.err != nil {
+		return it.err
 	}
+
+	return nil
 }
 
-func (it *encodedStepIterWithCollector) Next() bool {
-	if it.err != nil {
-		return false
+func (it *encodedStepIterWithCollector) nextSequential() error {
+	for i, iter := range it.seriesIters {
+		peek, collector, err := nextForStep(
+			it.seriesPeek[i],
+			iter,
+			it.seriesCollectors[i],
+			it.stepTime,
+		)
+
+		it.seriesPeek[i] = peek
+		it.seriesCollectors[i] = collector
+		if err != nil {
+			return err
+		}
 	}
 
-	bounds := it.meta.Bounds
-	if bounds.End().Before(it.stepTime) {
+	return nil
+}
+
+func (it *encodedStepIterWithCollector) Next() bool {
+	if it.err != nil || it.finished {
 		return false
 	}
 
-	for i := range it.seriesIters {
-		it.nextForStep(i, it.stepTime)
+	// NB: If no reader worker pool configured, use sequential iteration.
+	if it.workerPool == nil {
+		it.err = it.nextSequential()
+	} else {
+		it.err = it.nextParallel()
 	}
 
 	if it.err != nil {
 		return false
 	}
 
-	it.stepTime = it.stepTime.Add(bounds.StepSize)
-	return !bounds.End().Before(it.stepTime)
+	it.stepTime = it.stepTime.Add(it.meta.Bounds.StepSize)
+	next := !it.blockEnd.Before(it.stepTime)
+	if next && it.updateFn != nil {
+		it.updateFn()
```

Comment:
What's this used for btw?

Reply:
This collates the results post `Next()`, so when `Current()` is called on the iterator it does virtually no work. Also partially necessary here because the step collector `ConsolidateAndMoveToNext` functions currently mutate their internal steps, which is easy enough to change but this way tries to avoid the extra checks etc that would be added otherwise